### PR TITLE
feat: follow graphql-sse spec, fixing complete event & errors

### PR DESCRIPTION
**Pull request by Ancient-Dragon**
Pull request checklist
----

- [x] Please read our [contributor guide](https://github.com/Netflix/dgs-framework/blob/master/CONTRIBUTING.md)
- [x] Consider creating a discussion on the [discussion forum](https://github.com/Netflix/dgs-framework/discussions)
  first
- [x] Make sure the PR doesn't introduce backward compatibility issues
- [x] Make sure to have sufficient test cases

Pull Request type
----

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----
Change the event type on exception to data as per the spec.

On completing the publisher, send the complete event

Fixes Issue #1435

Alternatives considered
----
N/a

### CODE REVIEW

**Comment by Ancient-Dragon** [comment]
Hi, @srinivasankavitha please could I ask that someone in the netflix team reviews this PR?

**Comment by Ancient-Dragon** [comment]
Hey @srinivasankavitha any updates needed? Unforunately this PR is really needed for us since any errors in our subscriptions are not handled correctly (especially the PR to the spring-boot 2.7 branch)

**Comment by srinivasankavitha** [comment]
Sorry, we've been a bit swamped this week. I will try to get to this in the next day or so. Do you know if this could cause breaking changes?

**Comment by srinivasankavitha** [comment]
> Thanks @srinivasankavitha, I guess since this was broken and old versions of dgs weren't following the spec. If an up to date client were to call an old server that would error on a subscription then yes they would break. If both server + client were using the same version of dgs then there would be no breaking change IMO.

Thanks  for the context, @Ancient-Dragon. Just fyi, we are no longer back-porting major features, and especially breaking changes to the spring-boot-2.7 branch, since they will involve a major release, and the next major release (>6.x) is intended to be Spring Boot 3.x compatible. For maintenance reasons, we will only be selectively releasing spring-boot-2.7 but most new changes will be on the spring boot 3.x compatible master branch. 

If you do need spring-boot-2.7, I would recommend temporarily forking the repo and releasing for your own usage there. 

Since these changes will require a major version update in any case, we will be making this available in the next release 7.x, which will likely happen in 2 weeks.

I understand this is not convenient for your use case, but we do incur a significant overhead to  manage multiple versions without inadvertently breaking other users.

As mentioned, I'll finish reviewing this today but the changes will be available in the release a couple weeks from now. Thanks for your patience with this process.

**Comment by srinivasankavitha** [comment]
The original SSE implementation in the framework is based on `subscriptions-transport-sse` : https://github.com/CodeCommission/subscriptions-transport-sse . This was prior to the introduction of `graphql-sse` protocol described here: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md.

This seems to be the source of discrepancy. Your changes look good to me, but I think it would be better to introduce this as a new module to opt into call `dgs-subscriptions-graphql-sse`. This will prevent breakage of existing clients and also allow folks who want to use this newer standard SSE protocol to opt in. 

With that strategy, we should be able to avoid making this a breaking release. Let me know what you think.

File: graphql-dgs-subscriptions-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/sse/DgsSSESubscriptionHandler.kt
Comment (srinivasankavitha) [inline graphql-dgs-subscriptions-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/sse/DgsSSESubscriptionHandler.kt:123]: I think this is the main change that may lead to breaking existing clients. 

The original SSE implementation in the framework is based on subscriptions-transport-sse : https://github.com/CodeCommission/subscriptions-transport-sse . This was prior to the introduction of graphql-sse protocol described here: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md.

This seems to be the source of discrepancy. Your changes look good to me, but I think it would be better to introduce this as a new module to opt into call dgs-subscriptions-graphql-sse. This will prevent breakage of existing clients and also allow folks who want to use this newer standard SSE protocol to opt in.

With that strategy, we should be able to avoid making this a breaking release.

**Review by srinivasankavitha — CHANGES_REQUESTED** [review]
Added come comments and a proposal for moving forward. I'm traveling this next week, so will come back and resume.

File: graphql-dgs-subscriptions-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/sse/DgsSSESubscriptionHandler.kt
Comment (Ancient-Dragon) [reply graphql-dgs-subscriptions-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/sse/DgsSSESubscriptionHandler.kt]: I have made a new package called graphql-dgs-subscriptions-graphql-sse I have also created a brand new graphql client so that users can specifically choose which client they need depending on the server implementation

**Comment by Ancient-Dragon** [comment]
Based on the comments I have only updated the spring boot 3 branch, as you mentioned that we would be unable to get this kind of change into the spring boot 2.7 branch. If you feel that we can get this change into spring boot 2.7 then please let me know and I'll fix the other PR. Also once this is approved I can make the update to the docs PR :)

**Comment by Ancient-Dragon** [comment]
I don't seem to be able to re-run the check for building would someone from the dev team be able to run this for me?

**Comment by srinivasankavitha** [comment]
> I don't seem to be able to re-run the check for building would someone from the dev team be able to run this for me?

Done. A test in graphl client module seems to be failing though.

**Review by srinivasankavitha — APPROVED** [review]
The changes look good to me. Let's stick with adding this in the master (Spring Boot 3 compatible). I tested this out with `graphql-sse` client and seems to work as well.

The graphql-client tests are failing though. So we can merge this once that is fixed. The test failures occur locally as well, so easy to reproduce. If you are able to take a look and fix them that would be great. 

Thanks for all the work on this PR.

**Comment by Ancient-Dragon** [comment]
> Aside from minor rename suggestions, the changes look good to me. Let's stick with adding this in the master only and (Spring Boot 3 compatible) for now, and not the 2.7 branch. I tested this out with `graphql-sse` client and seems to work as well.
> 
> The graphql-client tests are failing though. So we can merge this once that is fixed. The test failures occur locally as well, so easy to reproduce. If you are able to take a look and fix them that would be great.
> 
> Thanks for all the work on this PR.

Interesting I don't get these failures locally :(

File: graphql-dgs-subscriptions-graphql-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/graphql/sse/DgsGraphqlSSESubscriptionHandler.kt
Comment (Ancient-Dragon) [reply graphql-dgs-subscriptions-graphql-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/graphql/sse/DgsGraphqlSSESubscriptionHandler.kt:1]: Done :)

File: graphql-dgs-subscriptions-graphql-sse-autoconfigure/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/graphql/sse/DgsGraphqlSSEAutoConfig.kt
Comment (Ancient-Dragon) [reply graphql-dgs-subscriptions-graphql-sse-autoconfigure/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/graphql/sse/DgsGraphqlSSEAutoConfig.kt:1]: Done :)

File: graphql-dgs-client/src/test/kotlin/com/netflix/graphql/dgs/client/GraphqlSSESubscriptionGraphQLClientTest.kt
Comment (Ancient-Dragon) [reply graphql-dgs-client/src/test/kotlin/com/netflix/graphql/dgs/client/GraphqlSSESubscriptionGraphQLClientTest.kt:36]: So the reason I had to exclude DgsSSEAutoConfig is because you can't have two rest controllers running on the same endpoint. And the reason I excluded WebSocketAutoConfig, is because the way I see these you a user wouldn't be mixing and matching subscription types. You either use websockets or graphql-sse or the old version of the protocol

Comment (srinivasankavitha) [reply graphql-dgs-client/src/test/kotlin/com/netflix/graphql/dgs/client/GraphqlSSESubscriptionGraphQLClientTest.kt:36]: Thanks, makes sense!

**Pull request merged**
